src/plugins/mer/mervirtualmachinedialog.cpp

```diff
@@ -0,0 +1,231 @@
+            if (!MerVirtualBoxManager::setVdiSize(i.key(), i.value()))
+                error.append(tr("Fail to change vdi (%1) size!\n").arg(i.key()));
+        }
+    }
+
+    if (error.isEmpty())
+        QDialog::accept();
+    else
+        QMessageBox::warning(this, vmName, error);
+}
+
+void MerVirtualMachineDialog::memorySizeChanged(int value)
+{
+    setBoldText(ui->memoryLabel, virtualMachineInfo.memorySizeMb != value);
+}
+
+void MerVirtualMachineDialog::cpuCountChanged(int value)
+{
+    setBoldText(ui->cpuLabel, virtualMachineInfo.cpuCount != value);
+}
+
+void MerVirtualMachineDialog::onVirtualMachineOffChanged(bool vmOff)
+{
+    // If the VM is started, cancel any unsaved changes
+    if (!vmOff) {
+        ui->memorySlider->setValue(virtualMachineInfo.memorySizeMb);
+        ui->cpuCountSlider->setValue(virtualMachineInfo.cpuCount);
+        vdisToResize.clear();
+        vdiCurrentChanged(ui->vdiComboBox->currentIndex());
+    }
+
+    ui->memorySlider->setEnabled(vmOff);
+    ui->memorySpinBox->setEnabled(vmOff);
+    ui->cpuCountSlider->setEnabled(vmOff);
+    ui->cpuCountSpinBox->setEnabled(vmOff);
+    ui->vdiSizeMbSlider->setEnabled(vmOff);
+    ui->vdiSizeGbSpinBox->setEnabled(vmOff);
+}
+
+void MerVirtualMachineDialog::setNewVdiSize(int index, int value)
+{
+    if (!isIndexValid(vdisInfo, index)) {
+        return;
+    }
+
+    const auto& currentVdi = vdisInfo.at(index);
+
+    if (currentVdi.capacity == value) {
+        if (vdisToResize.contains(currentVdi.path)) {
+            vdisToResize.remove(currentVdi.path);
+        }
+    }
+    else {
+        vdisToResize.insert(currentVdi.path, value);
+    }
+
+    setBoldText(ui->capacityLabel, currentVdi.capacity != value);
+}
+
+void MerVirtualMachineDialog::vdiSizeChanged(int value)
+{
+    setNewVdiSize(ui->vdiComboBox->currentIndex(), value);
+}
+
+void MerVirtualMachineDialog::vdiCurrentChanged(int index)
+{
+    if (!isIndexValid(vdisInfo, index)) {
+        return;
+    }
+
+    const auto& currentVdi = vdisInfo.at(index);
+    ui->capacityTextLabel->setText(QString("%1 Gb").arg(QString::number(currentVdi.capacity / 1024.0, 'f', 3)));
+
+    int maxStorageSizeMb = getAvailableStorageMb(currentVdi.path);
+
+    int vdiCapacity = currentVdi.capacity;
+    if (vdisToResize.contains(currentVdi.path)) {
+        vdiCapacity = vdisToResize.value(currentVdi.path);
+    }
+
+    // the minimum size is the current size because VBoxManager can't reduce storage size
+    ui->vdiSizeMbSlider->setRange(currentVdi.capacity, maxStorageSizeMb);
+    ui->vdiSizeGbSpinBox->setRange(currentVdi.capacity / 1024.0, maxStorageSizeMb / 1024.0);
+
+    ui->vdiSizeMbSlider->setValue(vdiCapacity);
+
+    ui->sizeOnDiskLabel->setText(QString("%1 Gb").arg(QString::number(currentVdi.sizeOnDisk / 1024.0, 'f', 3)));
+    ui->vdiPathEdit->setText(currentVdi.path);
+}
+
+int getTotalMemorySize()
+{
+    // default for now return 3.5GB
+    return 3584;
```

Comment:
If really need it, can do the calculation of the available PC memory

QT does not have its own implementation for this, it will be platform dependent code for win, OS X and linux